monitoring/collector/docker.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2017 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"net"
+	"net/http"
+	"net/url"
+
+	"github.com/gravitational/roundtrip"
+	"github.com/gravitational/trace"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	dockerSocketPath = "/var/run/docker.sock"
+	dockerURL        = "http://docker/"
+)
+
+// DockerCollector collect metrics about docker service status
+type DockerCollector struct {
+	client   *roundtrip.Client
+	dockerUp typedDesc
+}
+
+// NewDockerCollector returns initialized DockerCollector
+func NewDockerCollector() (*DockerCollector, error) {
+	transport := &http.Transport{
+		Dial: func(network, addr string) (net.Conn, error) {
+			return net.Dial("unix", dockerSocketPath)
+		},
+	}
+	client, err := NewRoundtripClient(dockerURL, roundtrip.HTTPClient(&http.Client{
+		Transport: transport,
+		Timeout:   collectMetricsTimeout,
+	}))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &DockerCollector{
+		client: client,
+		dockerUp: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "docker", "health"),
+			"Status of Docker daemon",
+			nil, nil,
+		), prometheus.GaugeValue},
+	}, nil
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+func (d *DockerCollector) Collect(ch chan<- prometheus.Metric) error {
+	var metric prometheus.Metric
+
+	healthy, err := d.healthStatus()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if healthy {
+		if metric, err = d.dockerUp.newConstMetric(1.0); err != nil {
+			return trace.Wrap(err, "failed to create prometheus metric: %v", err)
```

Comment:
i'd still get rid of this error handling if I could help it. It will almost never trigger (as long as it validates the cardinality of label sets) and moving this to init phase will make this code much cleaner